### PR TITLE
Fix compilation errors with Clang v16+. Also includes fixes for compiling with -std=c++23.

### PR DESCRIPTION
**Pull request by bdutro**
This PR fixes compilation errors I observed trying to compile with clang v16 and v17, as well as some errors from bumping the C++ standard to C++23.

### CODE REVIEW

File: sparta/sparta/pipeViewer/Outputter.hpp [hunk 1]
```diff
@@ -187,7 +187,7 @@ namespace sparta::pipeViewer
                     if(!dat.stringVector[i].empty()){
                         // We check if we have seen this exact pair, field and value before or not.
                         if(const auto& [val, str] = std::tie(dat.valueVector[i].first, dat.stringVector[i]);
-                           stringMap.try_emplace(std::make_tuple(dat.pairId, i, val), str).second) {
+                           stringMap.emplace(std::piecewise_construct, std::forward_as_tuple(dat.pairId, i, val), std::forward_as_tuple(str)).second) {
```
Comment (bdutro) [inline sparta/sparta/pipeViewer/Outputter.hpp:190]: Workaround for a Clang bug: https://github.com/llvm/llvm-project/issues/61415

This is essentially what `try_emplace` does under the hood, so performance shouldn't be affected.

File: sparta/sparta/resources/Buffer.hpp [hunk 2]
```diff
@@ -95,7 +95,7 @@ namespace sparta
          */
         struct DataPointer {
         private:
-            typename std::aligned_storage<sizeof(value_type), alignof(value_type)>::type object_memory_;
+            alignas(value_type) std::byte object_memory_[sizeof(value_type)];
```
Comment (bdutro) [inline sparta/sparta/resources/Buffer.hpp:98]: `std::aligned_storage` was deprecated in C++23: https://stackoverflow.com/questions/71828288/why-is-stdaligned-storage-to-be-deprecated-in-c23-and-what-to-use-instead

File: sparta/sparta/utils/Utils.hpp [hunk 3]
```diff
@@ -27,56 +27,56 @@
 /*!
  * \brief Custom literal for uint64
  */
-constexpr inline uint64_t operator "" _u64(unsigned long long i) {
+constexpr inline uint64_t operator ""_u64(unsigned long long i) {
```
Comment (bdutro) [inline sparta/sparta/utils/Utils.hpp:30]: Clang throws a compiler error if there is a space between the `""` and the `_`

File: sparta/src/StatisticsArchives.cpp [hunk 4]
```diff
@@ -196,6 +196,15 @@ void RootArchiveNode::saveTo(const std::string & dir)
...
 
+ArchiveNode::ArchiveNode(const std::string & name) :
+    name_(name)
+{
+}
+
+ArchiveNode::~ArchiveNode()
+{
+}
+
```
Comment (bdutro) [inline sparta/src/StatisticsArchives.cpp:207]: Moving these to the `.cpp` file helps avoid some pitfalls when you use an incomplete type with a smart pointer.

File: sparta/sparta/utils/SpartaSharedPointerAllocator.hpp
Comment (klingaard) [inline sparta/sparta/utils/SpartaSharedPointerAllocator.hpp:1]: Nice cleanup!

**Review by klingaard — APPROVED** [review]

**Pull request merged**